JENKINS-62572: Extend Downloadable API slightly

Zastai opened the pull request:
See [JENKINS-62572](https://issues.jenkins-ci.org/browse/JENKINS-62572).

This mainly adds:
- a static Downloadable.idFor(Class) method, which returns the ID to
  use for a given class
  - existing code using a class now uses this method to do the mapping
    to an ID value
- an overload of the static Downloadable.get() method taking a Class
  - applies idFor() and forwards to the existing String-based get()

It also adds a bit more JavaDoc to existing methods, and adds a few
@NonNull annotations.

No testcases added; the existing tests already exercise the code changes.

### Proposed changelog entries

* JENKINS-62572: Extend the DownloadService.Downloadable API slightly.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers


### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

Zastai committed:
JENKINS-62572: Extend Downloadable API slightly

This mainly adds:
- a static Downloadable.idFor(Class) method, which returns the ID to
  use for a given class
  - existing code using a class now uses this method to do the mapping
    to an ID value
- an overload of the static Downloadable.get() method taking a Class
  - applies idFor() and forwards to the existing String-based get()

It also adds a bit more JavaDoc to existing methods, and adds a few
@NonNull annotations.

daniel-beck commented:
I do not think this is a net positive change for multiple reasons.

The pattern to use a (simple) class name by default unless overridden is used in multiple places in Jenkins, and in none of them it would be expected that the class be anything but a reasonable default, and exposed as API like this. If you use a custom value, you'd make it a `public static final` field somewhere and reuse the value like that.

Additionally this introduces API that will not work as one would expect with any classes that customize the ID. There's plenty of APIs in Jenkins where you pass a `Class` as argument and get an instance of that class back, but they do not fail to find the instance just because it overrode a default.

Zastai commented:
The problem I ran into was that my downloadable auto-generated an ID from my class name, but then neither documents how that happens, nor provides a mapping method. That meant I could not reliably (from an API perspective - I know the actual mapping is unlikely to change) use `Downloadable.get()` to access it. I would have to use `all()` and test each using `instanceof`, or construct a throwaway instance to call `getId()`.

This tiny change brings API parity between the ways a Downloadable’s id can be set, and the ways you can retrieve the instance later. Making the mapping method public is optional; I’d have no real problem reducing its accessibility. But it does seem better to have it in one place instead of duplicating it.

daniel-beck commented:
TBH it looks like you want something like

```
        public static <T extends Downloadable> T get(Class<T> clazz) {
            for (Downloadable d : all()) {
                if(clazz == d.getClass())
                    return (T) d;
            }
            return null;
        }
```

This won't fail to find if IDs are overridden, and is just weird when two instances exist for the same `Downloadable` type for some reason – something the existing scheme supports.

Zastai commented:
But that would change the semantics of get(), whereas my version sticks to the existing ID-based lookup.
And given that the constructor taking a class does not put any restrictions on that class, it's not guaranteed that the downloadable returned by the `get()` will be an instance of that class.
For example, a plugin may continue to use a class for the ID for backwards compatibility, even if the actual implementing class has since changed.

Zastai commented:
In that sense, exposing and using that `idOf` (name easily changed) would be enough; then I can call that and use the existing `get()` for the lookup.

Zastai committed:
Use TODO, not FIXME, for @since.

oleg-nenashev commented on core/src/main/java/hudson/model/DownloadService.java:
```diff
@@ -410,8 +450,21 @@ public JSONObject reduce(List<JSONObject> jsonList) {
          * Returns the {@link Downloadable} that has the given ID.
+         *
+         * @param id The ID to look for.
          */
```
```suggestion
         */
         @CheckForNull
```

oleg-nenashev commented on core/src/main/java/hudson/model/DownloadService.java:
```diff
@@ -410,8 +450,21 @@ public JSONObject reduce(List<JSONObject> jsonList) {
+         *
+         * @since TODO
+         */
+        public static Downloadable get(@NonNull Class<?> clazz) {
```
```suggestion
        @CheckForNull
        public static Downloadable get(@NonNull Class<?> clazz) {
```

Zastai committed:
Add nullability annotations.

Zastai committed:
Merge branch 'master' into JENKINS-62572

oleg-nenashev reviewed (approved):
Looks good to me 👍

daniel-beck commented on core/src/main/java/hudson/model/DownloadService.java:
```diff
@@ -406,13 +449,29 @@ public JSONObject reduce(List<JSONObject> jsonList) {
         }
 
+        /**
+         * Returns the {@link Downloadable} that has an ID associated with the specified class.
```
This seems to lack an explanation of the ways this can go wrong. I don't think _ID associated with the specified class_ to someone unfamiliar with internals will make sense without more context.

Also e.g. `{@see #idFor(Class)}` could help.

Zastai replied:
Fair. I added a link to `idFor` from the JavaDoc of the relevant constructors too.

daniel-beck reviewed (commented):
Thank you for your explanation.

I'm still not confident this is a good idea, but you've addressed what problems I saw.

Zastai committed:
Improve JavaDoc.

Now the methods (constructors + get()) that generate an ID based on a
class, explicitly state that they use idFor() for that.

oleg-nenashev commented:
We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process
